pkg/planner/core/rule_outer_to_inner_join.go

```diff
@@ -0,0 +1,199 @@
+//  1. Traversal is preorder and the passed down predicate is checked for the left/right after join
+//  2. The ON clause and passed down predicate (from higher selects or joins) are comined and applied to join children.
+//     This logic depends on the join type with the following logic:
+//     - For left/right outer joins, the ON clause an be applied only on the inner side (null producing side)
+//     - For inner/semi joins, the ON clause can be applied on both children
+//     - For anti semi joins, ON clause applied only on left side
+//     - For all other cases, do not pass ON clause.
+func (*convertOuterToInnerJoin) optimize(_ context.Context, p base.LogicalPlan, _ *optimizetrace.LogicalOptimizeOp) (base.LogicalPlan, bool, error) {
+	planChanged := false
+	return p.ConvertOuterToInnerJoin(nil), planChanged, nil
+}
+
+// LogicalAggregation just works since schema = child + aggregate expressions. No need to map predicates.
+// Also, predicates involving aggregate expressions are not null filtering. IsNullReject always returns
+// false for those cases.
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *baseLogicalPlan) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self
+	for i, child := range p.Children() {
+		newChild := child.ConvertOuterToInnerJoin(predicates)
+		p.SetChild(i, newChild)
+	}
+	return p
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (p *LogicalJoin) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	innerTable := p.Children()[0]
+	outerTable := p.Children()[1]
+	switchChild := false
+
+	if p.JoinType == LeftOuterJoin {
+		innerTable, outerTable = outerTable, innerTable
+		switchChild = true
+	}
+
+	// First, simplify this join
+	if p.JoinType == LeftOuterJoin || p.JoinType == RightOuterJoin {
+		canBeSimplified := false
+		for _, expr := range predicates {
+			isOk := isNullFiltered(p.SCtx(), innerTable.Schema(), expr)
+			if isOk {
+				canBeSimplified = true
+				break
+			}
+		}
+		if canBeSimplified {
+			p.JoinType = InnerJoin
+		}
+	}
+
+	// Next simplify join children
+
+	combinedCond := mergeOnClausePredicates(p, predicates)
+	if p.JoinType == LeftOuterJoin || p.JoinType == RightOuterJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(combinedCond)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	} else if p.JoinType == InnerJoin || p.JoinType == SemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(combinedCond)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else if p.JoinType == AntiSemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	}
+
+	if switchChild {
```

Comment:
btw, why we need to swith child order ? It seems that the old logic in rule_predicate_push_down not include this part ?

Reply:
There is no child re-ordering here. The switch is just in the naming of innerchild and outerchild to have common code for left/right/inner. The flag is used to put the children in the same original order.